gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala

```diff
@@ -684,29 +754,54 @@ case class ColumnarOverrideRules(session: SparkSession)
     )
   }
 
-  override def preColumnarTransitions: Rule[SparkPlan] = plan =>
-    PhysicalPlanSelector.maybe(session, plan) {
-      setAdaptiveContext()
-      setOriginalPlan(plan)
-      transformPlan(preOverrides(), plan, "pre")
-    }
-
-  override def postColumnarTransitions: Rule[SparkPlan] = plan =>
-    PhysicalPlanSelector.maybe(session, plan) {
-      val planWithFallbackPolicy = transformPlan(fallbackPolicy(), plan, "fallback")
-      val finalPlan = planWithFallbackPolicy match {
-        case FallbackNode(fallbackPlan) =>
-          // we should use vanilla c2r rather than native c2r,
-          // and there should be no `GlutenPlan` any more,
-          // so skip the `postOverrides()`.
-          fallbackPlan
-        case plan =>
-          transformPlan(postOverrides(), plan, "post")
-      }
+  private def prepareFallback[T](plan: SparkPlan)(f: SparkPlan => T): T = {
+    setAdaptiveContext()
+    setOriginalPlan(plan)
+    try {
+      f(plan)
+    } finally {
       resetOriginalPlan()
       resetAdaptiveContext()
-      transformPlan(finallyRules(), finalPlan, "final")
     }
+  }
+
+  /**
+   * Note: Do not implement this API. We basically inject all of Gluten's physical rules through
+   * `postColumnarTransitions`.
+   *
+   * See: https://github.com/oap-project/gluten/pull/4790
+   */
+  final override def preColumnarTransitions: Rule[SparkPlan] = plan => {
+    // To infer caller's property: ApplyColumnarRulesAndInsertTransitions#outputsColumnar.
+    OutputsColumnarTester.wrap(plan)
+  }
+
+  override def postColumnarTransitions: Rule[SparkPlan] = plan => {
+    val outputsColumnar = OutputsColumnarTester.inferOutputsColumnar(plan)
+    val unwrapped = OutputsColumnarTester.unwrap(plan)
+    val vanillaPlan = ColumnarTransitions.insertTransitions(unwrapped, outputsColumnar)
+    withSuggestRules(suggestRules(outputsColumnar)).apply(vanillaPlan)
+  }
+
+  // Visible for testing.
+  def withSuggestRules(suggestRules: List[SparkSession => Rule[SparkPlan]]): Rule[SparkPlan] =
```

Comment:
How about naming `transformRules` ? The `suggest rules` is a bit hard to follow..

Reply:
I'll be good to both. I'd change to `withTransformRules` if it's confusing.